### ISSUE TITLE
IPVS backend active connections aren't distributed evenly

### ISSUE THREAD

**Issue by jslouisyou** [kubernetes/kubernetes#126378]
### What happened?

Hello Kubernetes Team,
I'm newbie about networks or `kube-proxy`, so I'd appreciate it if anyone could help me!

Recently I faced unbalanced network Tx traffic in all master nodes and I found that there were unbalance between the number of IPVS connection between `Services` exposed in k8s cluster and `kube-apiserver`.

This happened when any of `kube-apiserver` stopped or restarted, and `kube-proxy` seemed to re-established IPVS connections to alive `kube-apiserver` and this caused unbalance IPVS connections, in this case `MASTER2` was restarted, such as:
```
$ ipvsadm -ln | grep 10.233.0.1 -A 3
TCP  10.233.0.1:443 rr
  -> MASTER1:6443           Masq    1      13         0
  -> MASTER2:6443           Masq    1      2          0
  -> MASTER3:6443           Masq    1      12          0
```
as you can see, there is unbalance connection between all master nodes and this remains the save even after a long time.

Here's example of above situation in one of my cluster:
1. The number of IPVS backend connections still remains balanced with all `kube-apiserver` until then 00:30.
2. `kube-apiserver` restarted at 00:30, after then IPVS connections were changed and imbalance situation happened. This wasn't changed even if almost 16 hours elapsed.

![image](https://github.com/user-attachments/assets/3f0e2468-790f-4225-bf4a-ec2ed1031aed)

I'm wondering that:
* Why does this imbalance continue to persist?
* Is there a way to resolve this situation, such as modify `kube-proxy` configuration?

### What did you expect to happen?

When `kube-apiserver` stops:
1. `kube-proxy` will re-establish IPVS connections between `Service` and `kube-apiserver`.
4. `kube-proxy` will establish IPVS connection with alive `kube-apiserver`.

I expected that after `kube-apiserver` restarts and elapse several minutes and hours:
5. The number of IPVS connection in local node will be distributed evenly.

But IPVS connections are remaining same which I means that the number of IPVS connection is remaining also and there isn't significant changes after elapsed hours.

### How can we reproduce it (as minimally and precisely as possible)?

WIth deploying several `Services` (ensuring there are IPVS connections between `kube-apiserver` and node), kill or restart one of `kube-apiserver`.

### Anything else we need to know?

I'm using `IPVS` mode for `kube-proxy`. Here's the full configuration I'm using for `kube-proxy`:
```
apiVersion: v1
data:
  config.conf: |-
    apiVersion: kubeproxy.config.k8s.io/v1alpha1
    bindAddress: 0.0.0.0
    bindAddressHardFail: false
    clientConnection:
      acceptContentTypes: ""
      burst: 10
      contentType: application/vnd.kubernetes.protobuf
      kubeconfig: /var/lib/kube-proxy/kubeconfig.conf
      qps: 5
    clusterCIDR: 10.0.0.0/12
    configSyncPeriod: 15m0s
    conntrack:
      maxPerCore: 32768
      min: 131072
      tcpCloseWaitTimeout: 1h0m0s
      tcpEstablishedTimeout: 24h0m0s
    detectLocalMode: ""
    enableProfiling: false
    healthzBindAddress: 0.0.0.0:10256
    hostnameOverride: master01
    iptables:
      masqueradeAll: false
      masqueradeBit: 14
      minSyncPeriod: 0s
      syncPeriod: 30s
    ipvs:
      excludeCIDRs: []
      minSyncPeriod: 0s
      scheduler: rr
      strictARP: true
      syncPeriod: 30s
      tcpFinTimeout: 0s
      tcpTimeout: 0s
      udpTimeout: 0s
    kind: KubeProxyConfiguration
    metricsBindAddress: 0.0.0.0:10249
    mode: ipvs
    nodePortAddresses: []
    oomScoreAdj: -999
    portRange: ""
    showHiddenMetricsForVersion: ""
    udpIdleTimeout: 250ms
    winkernel:
      enableDSR: false
      networkName: ""
      sourceVip: ""
  kubeconfig.conf: |-
    apiVersion: v1
    kind: Config
    clusters:
    - cluster:
        certificate-authority: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
        server: https://127.0.0.1:6443
      name: default
    contexts:
    - context:
        cluster: default
        namespace: default
        user: default
      name: default
    current-context: default
    users:
    - name: default
      user:
        tokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
```

### Kubernetes version

K8S v1.21.6
K8S v1.28.3

### Cloud provider

On-premise K8S cluster provisioned via `kubespray`

### OS version

```
$ cat /etc/os-release 
NAME="Ubuntu"
VERSION="20.04.4 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.4 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal

$ uname -a
Linux n3-mgmt-master01 5.4.0-122-generic #138-Ubuntu SMP Wed Jun 22 15:00:31 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

```
$ crictl --version
crictl version v1.21.0
```

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by aojea** [comment]
by default communications to the apiserver use http2, that is a sticky protocol, it reuses the same TCP connection for all the requests.

Once your clients move to one apiserver they will remain there, you can use the flag `--goaway-chance` in the apiserver to improve this situation https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/

```

--goaway-chance float

  | To prevent HTTP/2 clients from getting stuck on a single apiserver, randomly close a connection (GOAWAY). The client's other in-flight requests won't be affected, and the client will reconnect, likely landing on a different apiserver after going through the load balancer again. This argument sets the fraction of requests that will be sent a GOAWAY. Clusters with single apiservers, or which don't use a load balancer, should NOT enable this. Min is 0 (off), Max is .02 (1/50 requests); .001 (1/1000) is a recommended starting point.


```

/kind support
/sig network

**Comment by jslouisyou** [comment]
Hi @aojea thank you so much for the quick response!
I'm going to apply this configuration and check whether the number of IPVS connections are changing or not.

BTW, it seems that client should need to handle `GOAWAY` in order to make new connection to other `kube-apiserver`. Am I right?

**Comment by aojea** [comment]
> BTW, it seems that client should need to handle `GOAWAY` in order to make new connection to other `kube-apiserver`. Am I right?

based on your description your clients are sticky because are using http2, only http2 clients will receive GOAWAYs frames.
Http1 client will not be sticky

**Comment by uablrek** [comment]
As @aojea has pointed out the TCP connections are not terminated and re-established, they are setup once and kept until something fails, or they are closed by the client. I am not experienced in http2, so my view is on TCP level, but the reasoning is independent of the upper layer protocol. This happens:

1. A number of clients connect to the service (clusterIP). They are about evenly distributed between the available servers (endpoints)
2. You restart one server (MASTER2)
3. The client connections are terminated, and they immediately re-connect to the Service
4. Now there are only 2 servers available, and the new client connections are distributed to those
5. The restarted server comes back, but the clients have already re-connected, so they are happy

This means after `5`, there are no connections to the re-started server. This condition will remain until all client terminate/re-connect the TCP connections. Hence if you ensure that the clients always re-connect, you will get an even distribution, **but...** that may be very inefficient because you re-establish the TCP connections all the time.

This phenomena is not `ipvs` specific, it will be the same for proxy-mode=iptables/nftables.

There is another potential problem that _is_ ipvs specific though: The ipvs instances on the K8s nodes are independent. This means that the _first time clients on different K8s nodes_ connect they will likely end up in the same server! This because the round-robin (rr) algorithm on the different nodes picks the first server independently of each other. This is however usually not a problem since it evens out after a short time, but you should be aware of it. Proxy-mode=iptables/nftables selects the server at random (no rr), so they will not suffer from this problem.

In any case, this is not something that can be fixed in K8s. The kube-proxy MUST NOT break established connections to enforce a uniform TCP-connection distribution for instance.

I close this issue since not a K8s problem.

/close

@shaneutt sorry to hijack this issue, hope you don't mind

**Comment by uablrek** [comment]
BTW, `kubectl get svc kubernetes -w` can be used to test client connects to the API-server

**Comment by jslouisyou** [comment]
Hi @uablrek

> There is another potential problem that is ipvs specific though: The ipvs instances on the K8s nodes are independent. This means that the first time clients on different K8s nodes connect they will likely end up in the same server! This because the round-robin (rr) algorithm on the different nodes picks the first server independently of each other.

Yes, as you mentioned in your comment, I also faced an abnormal behavior that established connection between `kube-apiserver` and service was following this order (I used `rr` algorithm for `ipvs` and there were 3 master nodes).

All connections were established firstly in `MASTER2` node and next for `MASTER1` and the last for `MASTER3` node (`MASTER1` -> `MASTER3` -> `MASTER1`). It caused imbalance connections but after elapsed some time, the connections were seemed to be distributed evenly.

And as @aojea mentioned, I applied `--goaway-chance` flag and after then I could find the number of IPVS connections are evenly distributed among all `kube-apiserver`. 

Thanks for your consideration!
